gold/nanomips.cc

```diff
@@ -1861,19 +1924,72 @@ class Target_nanomips : public Sized_target<size, big_endian>
   typedef typename elfcpp::Swap<size, big_endian>::Valtype Valtype;
   typedef typename elfcpp::Elf_types<size>::Elf_WXword Size_type;
 
+  struct Balc_trampoline
+  {
+    Address address;
+    Address target;
+    bool ignore{true};
+    bool is_trampoline{false};
+
+    Balc_trampoline(Address address_, Address target_)
+      : address(address_), target(target_) { }
+  };
+
+  struct Balc_trampoline_target
+  {
+    int count{0};
+    size_t first;
+    size_t trampoline;
+    size_t last;
+    Address target;
+  };
+
+  typedef std::vector<Balc_trampoline> Balc_trampoline_vector;
+
  public:
   Target_nanomips(const Target::Target_info* info = &nanomips_info)
     : Sized_target<size, big_endian>(info), state_(NO_TRANSFORM), got_(NULL),
       stubs_(NULL), rel_dyn_(NULL), copy_relocs_(elfcpp::R_NANOMIPS_COPY),
       gp_(NULL), attributes_section_data_(NULL), abiflags_(NULL),
-      got_mod_index_offset_(-1U), has_abiflags_section_(false)
+      got_mod_index_offset_(-1U), has_abiflags_section_(false),
+      done_with_trampolines_(false)
   { }
 
   // Make a new symbol table entry for the Nanomips target.
   Sized_symbol<size>*
   make_symbol(const char*, elfcpp::STT, Object*, unsigned int, uint64_t)
   { return new Nanomips_symbol<size>(); }
 
+  // Clear BALC trampolines.
+  void
+  clear_balc_trampolines()
+  { balc_trampolines_.clear(); }
+
+  inline bool is_generating_trampolines() const
+  { return state_ == TRAMPOLINE_B; }
+
+  // Add BALC trampoline
+  void
+  add_balc_trampoline(Address address, Address target)
+  {
+    balc_trampolines_.emplace_back(address, target);
+  }
+
+  // Find BALC trampoline by address
+  const Balc_trampoline*
+  find_balc_trampoline(Address address)
+  {
+    static size_t pos = 0;
+    size_t sz = balc_trampolines_.size();
+    for (size_t i = 0; i < sz; ++i)
```

Comment:
For future improvement: this requests would appear in a roughly ordered manner, so a sliding window would be ideal to limit this search.

Reply:
This has been addressed.